Classes/ios/Scanners/MTBBarcodeScanner.m

```diff
@@ -663,38 +664,61 @@ - (void)updateFocusPreferencesOfDevice:(AVCaptureDevice *)inputDevice reset:(BOO
     [inputDevice unlockForConfiguration];
 
     // this method will acquire its own lock
-    [self updateTorchModeForCurrentSettings];
+    [self updateForTorchMode:self.torchMode error:nil];
 }
 
 #pragma mark - Torch Control
 
 - (void)setTorchMode:(MTBTorchMode)torchMode {
-    _torchMode = torchMode;
-    [self updateTorchModeForCurrentSettings];
+    [self setTorchMode:torchMode error:nil];
+}
+
+- (BOOL)setTorchMode:(MTBTorchMode)torchMode error:(NSError **)error {
+    if ([self updateForTorchMode:torchMode error:error]) {
+        // we only update our internal state if setting the torch mode was successful
+        _torchMode = torchMode;
+        return YES;
+    }
+
+    return NO;
 }
 
 - (void)toggleTorch {
-    if (self.torchMode == MTBTorchModeAuto || self.torchMode == MTBTorchModeOff) {
-        self.torchMode = MTBTorchModeOn;
-    } else {
-        self.torchMode = MTBTorchModeOff;
+    switch (self.torchMode) {
+        case MTBTorchModeOn:
+            self.torchMode = MTBTorchModeOff;
+            break;
+
+        case MTBTorchModeOff:
+        case MTBTorchModeAuto:
+            self.torchMode = MTBTorchModeOn;
+            break;
     }
 }
 
-- (void)updateTorchModeForCurrentSettings {
+- (BOOL)updateForTorchMode:(MTBTorchMode)preferredTorchMode error:(NSError **)error {
     AVCaptureDevice *backCamera = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
-    if ([backCamera isTorchAvailable] && [backCamera isTorchModeSupported:AVCaptureTorchModeOn]) {
```

Comment:
I believe this should have always been `[backCamera isTorchModeSupported:[self avTorchModeForMTBTorchMode:self.torchMode]]`